xarray/conventions.py

```diff
@@ -476,18 +477,41 @@ def stackable(dim: Hashable) -> bool:
         if decode_coords == "all":
             for attr_name in CF_RELATED_DATA:
                 if attr_name in var_attrs:
-                    attr_val = var_attrs[attr_name]
-                    if attr_name not in CF_RELATED_DATA_NEEDS_PARSING:
-                        var_names = attr_val.split()
-                    else:
-                        roles_and_names = [
-                            role_or_name
-                            for part in attr_val.split(":")
-                            for role_or_name in part.split()
-                        ]
-                        if len(roles_and_names) % 2 == 1:
-                            emit_user_level_warning(f"Attribute {attr_name} malformed")
-                        var_names = roles_and_names[1::2]
+                    # fixes stray colon
+                    attr_val = var_attrs[attr_name].replace(" :", ":")
+                    var_names = attr_val.split()
+                    # if grid_mapping is a single string, do not enter here
+                    if (
+                        attr_name in CF_RELATED_DATA_NEEDS_PARSING
+                        and len(var_names) > 1
+                    ):
+                        # map the keys to list of strings
+                        roles_and_names = defaultdict(list)
+                        key = None
+                        for vname in var_names:
+                            if ":" in vname:
+                                key = vname.strip(":")
+                            else:
+                                if key is None:
+                                    raise ValueError(
+                                        f"First element {vname!r} of [{attr_val!r}] misses ':', "
+                                        f"cannot decode {attr_name!r}."
+                                    )
+                                roles_and_names[key].append(vname)
+                        # for grid_mapping keys are var_names
```

Comment:
man that's irritating :/

Reply:
Yes, for grid_mapping, we need the keys and for the other two the values. CF isn't always comprehensible. I split this into two iterations to make this more visible.

Reply:
I mean, I can split this if necessary for visibility. WDYT?

Reply:
the comment helps, thanks!